File: basics/go-programs/08_basic_server.go
package main

import (
	"fmt"
	"net/http"
)

/*
	GO PROGRAM 8: BASIC HTTP SERVER
	Concepts:
	- 'net/http' is Go's powerful standard library for web servers.
	- 'http.HandleFunc()' registers a function to handle a specific URL path.
	- 'http.ListenAndServe()' starts the server on a given port.
*/

func homeHandler(w http.ResponseWriter, r *http.Request) {
	// Send a plain text response
	fmt.Fprintf(w, "Welcome to the Go Home Page!")
}

func helloHandler(w http.ResponseWriter, r *http.Request) {
	// Extract a query parameter (e.g., /hello?name=Gopher)
	name := r.URL.Query().Get("name")
	if name == "" {
		name = "Guest"
	}
	fmt.Fprintf(w, "Hello, %s!", name)
}

func main() {
	// Define routes
	http.HandleFunc("/", homeHandler)
	http.HandleFunc("/hello", helloHandler)
	
	// Start the server
	fmt.Println("Server starting on port 8080...")
	fmt.Println("Try visiting http://localhost:8080 or http://localhost:8080/hello?name=YourName")
	
	// Listen and serve on port 8080
	err := http.ListenAndServe(":8080", nil)
	if err != nil {
		fmt.Printf("Error starting server: %v\n", err)
	}
}

File: basics/node-programs/08_basic_server.js
const http = require('http');
const url = require('url');

/*
	NODE.JS PROGRAM 8: BASIC HTTP SERVER
	Concepts:
	- Using the built-in 'http' module.
	- Handling 'request' and 'response'.
	- Parsing URLs and query parameters.
*/

const server = http.createServer((req, res) => {
	// Parse the URL to get the path and query
	const parsedUrl = url.parse(req.url, true);
	
	// Set the response header
	res.setHeader('Content-Type', 'text/plain');
	
	if (parsedUrl.pathname === '/') {
		// Home logic
		res.end("Welcome to the Node.js Home Page!");
	} else if (parsedUrl.pathname === '/hello') {
		// Extract query parameter (e.g., /hello?name=Coder)
		const name = parsedUrl.query.name || "Guest";
		res.end(`Hello, ${name}!`);
	} else {
		res.statusCode = 404;
		res.end("Page not found");
	}
});

const PORT = 3000;
server.listen(PORT, () => {
	console.log(`Server starting on http://localhost:${PORT}...`);
	console.log(`Try visiting http://localhost:${PORT} or http://localhost:${PORT}/hello?name=YourName`);
});
